mountpoint-s3-crt/src/auth/imds_client.rs

```diff
@@ -0,0 +1,126 @@
+//! A client for retrieving ec2 instance metadata.
+
+use crate::auth::auth_library_init;
+use crate::common::allocator::Allocator;
+use crate::common::error::Error;
+use crate::io::channel_bootstrap::ClientBootstrap;
+use crate::io::retry_strategy::RetryStrategy;
+use crate::CrtError;
+use mountpoint_s3_crt_sys::{
+    aws_byte_buf, aws_imds_client, aws_imds_client_get_instance_type, aws_imds_client_new, aws_imds_client_options,
+    aws_imds_client_release,
+};
+use std::ffi::{OsStr, OsString};
+use std::os::unix::prelude::OsStrExt;
+use std::ptr::NonNull;
+
+/// A client for instance metadata query.
+#[derive(Debug)]
+pub struct ImdsClient {
+    /// A pointer to the underlying `aws_imds_client`
+    inner: NonNull<aws_imds_client>,
+}
+
+///Configurations for creating a new [ImdsClient].
+#[derive(Debug, Default)]
+pub struct ImdsClientConfig {
+    /// The structure that can pass into the CRT's functions.
+    inner: aws_imds_client_options,
+
+    /// The [ClientBootstrap] to use to create connection to IMDS.
+    client_bootstrap: Option<ClientBootstrap>,
+
+    /// The [RetryStrategy] to use to reschedule failed requests to IMDS. This is reference counted,
+    /// so we only need to hold onto it until this [ImdsClientConfig] is consumed, at which point the
+    /// client will take ownership.
+    retry_strategy: Option<RetryStrategy>,
+}
+
+impl ImdsClient {
+    /// Create a new [ImdsClient]
+    pub fn new(allocator: &Allocator, config: ImdsClientConfig) -> Result<Self, Error> {
+        auth_library_init(allocator);
+
+        // SAFETY: `config` is moved into the [ImdsClient] on success, so `config.inner` (and the values
+        // inside) are guaranteed to live at least as long as this [ImdsClient] does. `allocator` is
+        // guaranteed to be a valid allocator because of the type-safe wrapper.
+        let inner = unsafe { aws_imds_client_new(allocator.inner.as_ptr(), &config.inner).ok_or_last_error()? };
+
+        Ok(Self { inner })
+    }
+
+    /// Make a get instance type request to ec2 hosts using [ImdsClient].
+    pub fn get_instance_type<F>(&self, callback: F) -> Result<i32, Error>
+    where
+        F: FnOnce(Result<OsString, Error>) + 'static,
+    {
+        let callback_closure: OnGetResource = Box::new(callback);
+        let callback_wrapper: Box<ImdsClientGetResourceCallback> =
+            Box::new(ImdsClientGetResourceCallback(Box::new(callback_closure)));
+        let callback_wrapper: &mut ImdsClientGetResourceCallback = Box::leak(callback_wrapper);
+        // callback_wrapper is safe so the raw pointer referring to that is safe as well.
+        let callback_raw_ptr: *mut libc::c_void =
+            callback_wrapper as *mut ImdsClientGetResourceCallback as *mut libc::c_void;
+        let get_resource_callback_fn_ptr: Option<unsafe extern "C" fn(*const aws_byte_buf, i32, *mut libc::c_void)> =
+            Some(imds_client_get_resource_callback);
```

Comment:
I think we can probably reorganise this a bit to make it a bit clearer.

Is the callback_closure double boxed? Does it need to be?

I think we can use `Box::into_raw(T) -> *mut T` instead to simplify here too. This SO answer describes our use case: https://stackoverflow.com/a/73330433

Dropping some of the types may also help make the code a bit clearer.

```suggestion
        let callback_closure: OnGetResource = Box::new(callback);
        let callback_wrapper: Box<ImdsClientGetResourceCallback> =
            Box::new(ImdsClientGetResourceCallback(callback_closure));
        // callback_wrapper is safe so the raw pointer referring to that is safe as well.
        let callback_raw_ptr = Box::into_raw(callback_wrapper) as *mut libc::c_void;
        let get_resource_callback_fn_ptr = Some(imds_client_get_resource_callback);
```

Reply:
Will fix, and that will simplify type conversion